api/services/registry/server.go

```diff
@@ -106,3 +109,38 @@ func (s *Server) Proxy(stream Registry_ProxyServer) error {
 
 	return nil
 }
+
+// CopyWithDeadline copies data from a net.Conn using a read deadline. The
+// process will fail with a timeout error if no data is read for the defined
+// period.
+func CopyWithDeadline(conn net.Conn, w io.Writer) (int64, error) {
+	var t int64
+	for {
+		err := conn.SetReadDeadline(time.Now().Add(readDeadline))
+		if err != nil {
+			return 0, err
+		}
+		buf := make([]byte, 32*1024)
```

Comment:
can we move this outside of the for loop? reallocating this buffer each time might be expensive.